### DISCUSSION TITLE
How to programmatically add fields to a Data Model

### DISCUSSION THREAD

**Discussion by ched-dev** [directus/directus#12368] [Archive]
> tl;dr Looking for a way to programmatically add fields to a Data Model

I made a Hook Extension for [attaching EXIF data to a Data Model](https://github.com/ched-dev/directus-extension-exif-attacher) and now I'm looking into automating the addition of the fields to the Data Model.

I've looked into migrations, but not sure I want to make the changes directly in knex when it really should be done in the manner the UI would add fields. 

I'm thinking of creating a Module that can create a Data Model with all the required fields (or add fields to an existing Data Model), but worried I will go down the path of re-creating lots of the Content Module functionality. 

I'm currently looking into @alejandro-tss's [schema-builder-kit](https://github.com/directus-community/schema-builder-kit) but are there any other suggest solutions?

**Comment by rijkvanzanten** [comment]
You could potentially rely on the schema migration snapshot structure, or use the /collections, /fields, and /relations endpoints directly[^1].

[^1]: That's basically what `schema-builder-kit` does under the hood

**Reply by ched-dev** [reply]
Didn't think about using the endpoints directly... That's a good solution too. The `schema-builder-kit` is a nice abstraction to, with auth built in.  I will look deeper into the two options.

Do snapshots copy the whole apps schema or can I create one for specific collection(s)?

**Reply by rijkvanzanten** [reply]
Snapshots right now snapshot the whole schema, so it isn't really the best tool for the job to apply "snippets" of schema

**Reply by alexbrowng** [reply]
The snapshot system is a good utility, to support the load of "snippets" you should be able to mark which actions you want to apply. Currently, it creates new records, updates existing ones and deletes non-existing ones, if you could choose, create new, update or delete or a combination, e.g. create and update but not delete, then partial snapshots could be created.

**Comment by alexbrowng** [comment]
As for the schema builder, I have not been able to continue with the development, although I have a version halfway in local that supports the management of presets, roles and permissions and I would like to allow the loading of initial data, for example, languages, categories, users, images, etc., but I don't know when I can resume it.

**Reply by ricricucit** [reply]
@rijkvanzanten is there any plan, from the directus core team, to take this on or to offer alternatives to it?

**Reply by rijkvanzanten** [reply]
Absolutely! This is being included in the work for https://github.com/directus/directus/discussions/13041
